### DISCUSSION TITLE
Vertical wind profile

### DISCUSSION THREAD

**Discussion by MiguelMarante** [NatLabRockies/floris#655] [Q&A] [answered]
Hello,

I would like to ask why does the vertical wind profile given by floris is not the same one would get using the power law. I used the same shear exponent (0.2) and hub height (90), however the results are not the same.

Am I overlooking something?

Best regards,
Miguel

```python
import numpy as np

from floris.tools import FlorisInterface

import matplotlib.pyplot as plt



# Initialize FLORIS with the given input file via FlorisInterface.
# For basic usage, FlorisInterface provides a simplified and expressive
# entry point to the simulation routines.
fi = FlorisInterface("inputs/gch.yaml")

nelx = 101

solver_settings = {
    "type": "turbine_grid",
    "turbine_grid_points": nelx
}

# Create a 4-turbine layouts
fi.reinitialize(layout_x=[0], layout_y=[0], solver_settings=solver_settings)


yaw_angles = np.array([[[0]]])
# Calculate wake
fi.calculate_wake(yaw_angles = yaw_angles)

# Collect the wind speed at all the turbine points
u_points = fi.floris.flow_field.u

D = 126
hubHt = 90

xaxis = D/nelx * np.arange(0,nelx) + hubHt - D/2 + D/nelx/2
analytical_profile = 8 * ((xaxis)/hubHt)**0.2
```

![image](https://github.com/NREL/floris/assets/133772056/66b5019b-512f-4be7-8500-0ea6d42fe1ab)

**Comment by paulf81** [comment]
Hi I think the issue is just the shear exponent in set to 0.12 inputs/gch.yaml (line 126), I ran this code (using the new point grabbing code in v3.4) and the curves match:

```
fi = FlorisInterface("inputs/gch.yaml")

nelx = 101
D = 126.
hubHt = 90.

z = D/nelx * np.arange(0,nelx) + hubHt - D/2 + D/nelx/2
analytical_profile = 8 * ((z)/hubHt)**.12

# Collect same points from FLORIS
fi.reinitialize(layout_x=[1000],layout_y=[0], wind_speeds=[8.0]) # Put the only turbine out of the way

points_z = z
points_x = np.zeros_like(z)
points_y = np.zeros_like(z)

u_at_points = fi.sample_flow_at_points(points_x, points_y, points_z)


fig, ax = plt.subplots()
ax.plot(analytical_profile,z)
ax.plot(u_at_points.squeeze(),z,'r--')
ax.grid(True)

plt.show()

```

**Reply by MiguelMarante** [reply]
Problem solved!! 

It was not the exponent that was poorly defined, it was the use of another function:

![image](https://github.com/NREL/floris/assets/133772056/98fb6729-14df-4bdd-aeb5-887bdd3c5726)

I am not sure about the coordinates of the wind speeds given by this function, but the "sample_flow_at_points" function does the trick.

Thank you so much!!

**Reply by paulf81** [reply]
Very glad to hear it, thank you!

**Comment by rafmudaf** [comment, marked as answer]
@MiguelMarante thanks for posting your question and apologies for the super late response.
In your script, you're comparing the shear profile at different heights. Specifically, you're plotting the vertical locations across the full span of the rotor swept area, but FLORIS fits a square-grid into a smaller portion of the rotor swept area. See https://github.com/NREL/floris/discussions/502#discussioncomment-3691865 for a diagram of the FLORIS grid relative to the swept area. The setting for how large to make the grid relative to the rotor diameter was a user setting in floris v2, but it is now [hard-coded to half the radius](https://github.com/NREL/floris/blob/v3.4.1/floris/simulation/grid.py#L229). Additionally, https://github.com/NREL/floris/issues/499 provides some info on the impact of the the size of the grid as well as the number of points in it, https://github.com/NREL/floris/pull/649 partially addresses this by adding a new grid and integration method, and upcoming work will make the new grid the default.

Back to your script, if you plot the curves at the appropriate heights, they both line up. The plot below shows your shear profile and the one from FLORIS. The code for this is given below the plot.

Also, just a heads up that it's very helpful in general to share the actual text as well as the full script you used to generate results. Sometimes the problem is a plotting issue, for example, and it's helpful for project maintainers to see how you created the results that you're showing. Including the actual text means that I can copy/paste it directly to try to recreate what you're seeing. I replaced your screen shot in the OP with the text.

![shear](https://github.com/NREL/floris/assets/13797903/7e0d0246-b239-49c9-a057-ac32822da814)

```python
import matplotlib.pyplot as plt
import numpy as np
from floris.tools import FlorisInterface

nelx = 7
solver_settings = {
    "type": "turbine_grid",
    "turbine_grid_points": nelx
}
fi = FlorisInterface("inputs/gch.yaml")

fi.reinitialize(layout_x=[0.0], layout_y=[0.0], solver_settings=solver_settings)
fi.calculate_wake()

grid_heights = fi.floris.grid.z_sorted[0,0,0,0]
print("FLORIS grid locations - z:")
print(grid_heights)

print("FLORIS velocities at z-locations:")
u_points = fi.floris.flow_field.u[0,0,0,0]
print(u_points)

print("---")

D = 126.0
hubHt = 90.0
radius = D/2.0
SHEAR = 0.12

xaxis = D/nelx * np.arange(0,nelx) + hubHt - D/2 + D/nelx/2  # this should probably be "yaxis"
print("@MiguelMarante grid locations - z:")
print(xaxis)

analytical_profile = 8 * (xaxis / hubHt) ** SHEAR
print("@MiguelMarante analytical shear profile at z-locations:")
print(analytical_profile)

plt.plot(u_points, grid_heights, marker="+", label="FLORIS")
plt.plot(analytical_profile, xaxis, label="@MiguelMarante")
plt.legend()
plt.grid()
plt.show()
```

**Reply by rafmudaf** [reply]
Also @MiguelMarante 101 points in the grid is way too much. I think you were doing that to get a smooth plot, but just in case your simulations are going slow, double check that you aren't using 101 grid points.

**Reply by rafmudaf** [reply]
Here's the output of the script above:

```
FLORIS grid locations - z:
[ 58.5  69.   79.5  90.  100.5 111.  121.5]
FLORIS velocities at z-locations:
[7.59695563 7.74894854 7.88179148 8.         8.10663862 8.2038865
 8.29335086]
---
@MiguelMarante grid locations - z:
[ 36.  54.  72.  90. 108. 126. 144.]
@MiguelMarante analytical shear profile at z-locations:
[7.16699639 7.52433524 7.78862484 8.         8.17695743 8.32962311
 8.46417015]
```

**Reply by rafmudaf** [reply]
@paulf81 @MiguelMarante I'm going to mark this as the accepted answer, but please comment if this is incorrect.

**Reply by MiguelMarante** [reply]
Hello Rafael,

Thank you very much for your response, it was very enlightening. I just have one question left: Are we considering only half the radius because otherwise there would be points outside the rotor swept area?

Yeah, I apologise for not including the full script. I am new to programming and also to github.

Thank you very much for all your input.

Best regards,
Miguel Pereira

**Reply by rafmudaf** [reply]
@MiguelMarante Initially (a few years ago, at this point), the square grid covered the entire swept area and then we filtered out points that were outside of the circle. Then, we decided to move all the points inside the circle, but we were putting the corners of the square grid on the edge of the swept area circle. To be honest, I don't remember why but eventually we settled on the 1/2 radius setting.
That being said, I'm working on validating a new grid and integration method that will move beyond the square grid and will have a better convergence rate relative to the number of grid points. It's already available in FLORIS (see the pull requests I linked to above), and it'll be the default soon.
